### PR TITLE
Refactor block syncing

### PR DESCRIPTION
**Pull request by NullSoldier**
This is a placeholder PR to fix the current syncing issues were seeing. I'll squash this and reorganize the commits, squash, and split them up as I get closer to a finished version.

**TODO**
- [x] Add GetBlockHashesRequest and GetBlockHashesResponse
- [x] Add GetBlocksRequest and GetBlocksResponse
- [x] Find ancestor using linear search
- [x] Find ancestor using binary search
- [x] Sync pagination from peer using ancestor
- [x] Handle peer disconnection
- [x] Handle abort syncing when node shutting down
- [x] Ignore gossip messages when syncing
- [x] Handle gossip message that becomes orphan (start syncing)
- [x] Get tests passing
- [x] BUG: Assert peer.head
- [ ] ~~BUG: Invalid GetBlocksResponse (cannotSatisfy) from other person~~
- [ ] ~~BUG: maintainOneConnectionPerPeer() should not run while pending RPC~~

**Ideas**
1. Only sync left to right
1. Send head, work in identity message
1. Sync from peer with higher work than ours
1. Find common ancestor block with peer before syncing using binary search (similar to etherium)
1. Start syncing from common ancestor block

### CODE REVIEW

File: ironfish/src/utils/error.ts [hunk 1]
```diff
@@ -3,14 +3,26 @@
...
 }
 
+/**
+ * This is used to unwrap a message from an error
+ *
+ * Falls back to JSON.stringify the error if we cannot get the message
+ */
+export function renderError(error: unknown, stack = false): string {
+  if (stack && error instanceof Error && error.stack) {
+    return error.stack
```
[the remark this answers is not in the document]
Comment (NullSoldier) [reply ironfish/src/utils/error.ts:22]: Stack in javascript has the error itself

```javascript
Uncaught Error: This is an error
    at crash (REPL4:1:26)
```

File: ironfish/src/consensus/consensus.ts [hunk 2]
```diff
@@ -36,3 +36,13 @@ export const IRON_FISH_YEAR_IN_BLOCKS = 2100000
...
+
+/**
+ * The maximum allowed requested blocks by the network
+ */
+export const MAX_REQUESTED_BLOCKS = 50
+
+/**
+ * The network has a target block time of 15 seconds
+ */
+export const TARGET_BLOCK_TIME_MS = 15 * 1000
```
[the remark this answers is not in the document]
Comment (NullSoldier) [reply ironfish/src/consensus/consensus.ts:48]: Well, I'm trying to move all network parameters into one place so they can be reused, and changed per network.

Like, `IRON_FISH_YEAR_IN_BLOCKS` technically uses `TARGET_BLOCK_TIME_MS` to calculate itself, and all these numbers can change between networks.

I don't know why it's not used in target, but it should be. I'll have to ask Elena that.

File: ironfish/src/network/messages.ts [hunk 3]
```diff
@@ -195,6 +193,7 @@ export function isPeerList(obj: unknown): obj is PeerList {
 export enum DisconnectingReason {
   ShuttingDown = 0,
   Congested = 1,
+  Unknown = 2,
```
[the remark this answers is not in the document]
Comment (NullSoldier) [reply ironfish/src/network/messages.ts:196]: It's for future proofing, so we don't have to break the protocol, but I can remover it.

File: ironfish/src/network/messages.ts [hunk 4]
```diff
@@ -349,12 +350,107 @@ export type BlockRequest = Message<
   }
 >
 
-/**
- * Type narrowing to verify that the block has the hash and sequence parameters
- * and that they are either undefined or strings.
- */
-export function isBlockRequestPayload(obj: PayloadType): obj is BlockRequest['payload'] {
-  return obj != null && 'hash' in obj && (typeof obj.hash === 'string' || obj.hash === null)
```
[the remark this answers is not in the document]
Comment (NullSoldier) [reply ironfish/src/network/messages.ts:357]: Deleted.

File: ironfish/src/network/messages.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (NullSoldier) [reply ironfish/src/network/messages.ts]: removed

File: ironfish/src/consensus/consensus.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (dguenther) [reply ironfish/src/consensus/consensus.ts:48]: oh also, the name of this ends in `_MS` but the unit is actually seconds, I think MAX_SYNCED_AGE_MS has the same issue

**Review by dguenther — APPROVED** [review]

**Pull request merged**